outlines/models/llamacpp.py

```diff
@@ -1,131 +1,170 @@
-            for token_pos, token_id in enumerate(seq):
-                if token_id == self.pad_token_id:
-                    break
-                batch.token[batch.n_tokens] = token_id.item()
-                batch.pos[batch.n_tokens] = token_pos
-                batch.seq_id[batch.n_tokens][0] = seq_idx
-                batch.n_seq_id[batch.n_tokens] = 1
-                batch.logits[batch.n_tokens] = False
+        mask = torch.full((scores.shape[-1],), -math.inf, device="cpu").numpy()
+        mask[allowed_tokens] = 0
+        biased_scores = scores + mask
 
-                batch.n_tokens += 1
-                self.n_past += 1
+        biased_scores[self.tokenizer.eos_token_id] = 0
 
-            batch.logits[batch.n_tokens - 1] = True
-            seq_token_ids.append(batch.n_tokens - 1)
+        return biased_scores
 
-        if self.llama_instance.llama_decode(self.ctx, batch) != 0:
-            print("Error decoding")
 
-        all_logits = []
-        for seq_token in seq_token_ids:
-            logits = self.llama_instance.llama_get_logits_ith(self.ctx, seq_token)
-            logits_list = (ctypes.c_float * self.n_vocab)(
-                *[logits[token_id] for token_id in range(self.n_vocab)]
-            )
-            logits_tensor = torch.tensor(logits_list)
-            all_logits.append(logits_tensor)
+class JSONLogitsProcessor(RegexLogitsProcessor):
+    def __init__(self, schema, llm):
+        """Compile the FSM that drives the JSON-guided generation.
+
+        Parameters
+        ----------
+        schema
+            A JSON schema that encodes the structure we want the model to generate
+        llm
+            An instance of `vllm.LLM`
+
+        """
+        if isinstance(schema, dict):
+            schema = json.dumps(schema)
+        regex_string = build_regex_from_object(schema)
+        super().__init__(regex_string, llm)
+
 
-        self.llama_instance.llama_batch_free(batch)
+class LlamaCpp(Llama):
+    """Represents a `llama_cpp` model."""
+
+    def __init__(self, model, **kwargs):
+        super().__init__(model, **kwargs)
+
+        self.device = "cpu"
 
-        stacked_logits = torch.stack(all_logits)
-        return stacked_logits, None
+        self.tokenizer = LlamaCppTokenizer(self)
 
     def __call__(
         self,
         input_ids: torch.LongTensor,
         attention_mask: torch.LongTensor,
         past_key_values: Optional[Tuple] = None,
     ) -> torch.FloatTensor:
-        logits, kv_cache = self.forward(input_ids, attention_mask, past_key_values)
-        next_token_logits = logits
+        all_logits = []
+
+        for i in range(input_ids.size(0)):
+            super().eval(input_ids[i, self.n_tokens :])
+
+            logits = super().eval_logits
+            logits_tensor = torch.FloatTensor(list(logits))
+            all_logits.append(logits_tensor[0, :])
 
-        return next_token_logits, kv_cache
+        return torch.stack(all_logits), None
+
+    def sample(
```

Comment:
We don't need to subclass `sample` until we make `Llama` support other sampling algorithms)